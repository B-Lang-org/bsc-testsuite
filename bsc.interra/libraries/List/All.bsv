package All;

import List :: *;
function Bool f (Int #(32) a);
     if (a < 0)
        return False;
     else
        return True;
endfunction


module mkTestbench_All();
   List #(Int #(32)) my_list1 = Cons (0, Cons (-1, Cons (2, Cons (-3, Cons (4, Nil)))));
   List #(Int #(32)) my_list2 = Cons (0, Cons (2, Cons (4, Nil)));



   rule fire_once (True);
      if (all(f, my_list1) != False || all(f, my_list2) != True) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_All
endpackage : All
