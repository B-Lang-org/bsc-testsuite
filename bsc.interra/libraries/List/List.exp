test_c_veri_bsv_multi All mkTestbench_All {} "mkTestbench_All.out.expected" 

if {$ctest == 1} {
compile_object_fail And.bsv mkTestbench_And
compare_file And.bsv.bsc-ccomp-out
}
#link_objects_pass {"mkTestbench_And"} mkTestbench_And
#sim_final_state mkTestbench_And 10010
#compare_file "mkTestbench_And.final-state"

test_c_veri_bsv_multi Any mkTestbench_Any {} "mkTestbench_Any.out.expected" 

test_c_veri_bsv_multi Append mkTestbench_Append {} "mkTestbench_Append.out.expected" 

test_c_veri_bsv_multi Concat mkTestbench_Concat {} "mkTestbench_Concat.out.expected" 

test_c_veri_bsv_multi Drop mkTestbench_Drop {} "mkTestbench_Drop.out.expected" 

test_c_veri_bsv_multi DropWhile mkTestbench_DropWhile {} "mkTestbench_DropWhile.out.expected" 

test_c_veri_bsv_multi DropWhileRev mkTestbench_DropWhileRev {} "mkTestbench_DropWhileRev.out.expected" 

test_c_veri_bsv_multi Elem mkTestbench_Elem {} "mkTestbench_Elem.out.expected" 

test_c_veri_bsv_multi Equality mkTestbench_Equality {} "mkTestbench_Equality.out.expected" 

test_c_veri_bsv_multi Filter mkTestbench_Filter {} "mkTestbench_Filter.out.expected" 

test_c_veri_bsv_multi Fold mkTestbench_Fold {} "mkTestbench_Fold.out.expected" 

test_c_veri_bsv_multi Foldl mkTestbench_Foldl {} "mkTestbench_Foldl.out.expected" 

test_c_veri_bsv_multi Foldl1 mkTestbench_Foldl1 {} "mkTestbench_Foldl1.out.expected" 

test_c_veri_bsv_multi Foldn mkTestbench_Foldn {} "mkTestbench_Foldn.out.expected" 

test_c_veri_bsv_multi Foldr mkTestbench_Foldr {} "mkTestbench_Foldr.out.expected" 

test_c_veri_bsv_multi Foldr1 mkTestbench_Foldr1 {} "mkTestbench_Foldr1.out.expected" 

test_c_veri_bsv_multi Head mkTestbench_Head {} "mkTestbench_Head.out.expected" 

test_c_veri_bsv_multi Init mkTestbench_Init {} "mkTestbench_Init.out.expected" 

test_c_veri_bsv_multi JoinActions mkTestbench_JoinActions {} "mkTestbench_JoinActions.out.expected" 

test_c_veri_bsv_multi JoinRules mkTestbench_JoinRules {} "mkTestbench_JoinRules.out.expected" 

test_c_veri_bsv_multi Last mkTestbench_Last {} "mkTestbench_Last.out.expected" 

test_c_veri_bsv_multi Length mkTestbench_Length {} "mkTestbench_Length.out.expected" 

test_c_veri_bsv_multi Map mkTestbench_Map {} "mkTestbench_Map.out.expected" 

test_c_veri_bsv_multi MapAccumL mkTestbench_MapAccumL {} "mkTestbench_MapAccumL.out.expected" 

test_c_veri_bsv_multi MapAccumR mkTestbench_MapAccumR {} "mkTestbench_MapAccumR.out.expected" 

test_c_veri_bsv_multi MapPairs mkTestbench_MapPairs {} "mkTestbench_MapPairs.out.expected" 

test_c_veri_bsv_multi Mapn mkTestbench_Mapn {} "mkTestbench_Mapn.out.expected" 

if {$ctest == 1} {
compile_object_fail Null.bsv mkTestbench_Null
compare_file Null.bsv.bsc-ccomp-out
}
#link_objects_pass {"mkTestbench_Null"} mkTestbench_Null
#sim_final_state mkTestbench_Null 10010
#compare_file "mkTestbench_Null.final-state"

compile_verilog_pass OneHotSelect.bsv mkTestbench_OneHotSelect

if {$ctest == 1} {
compile_object_fail Or.bsv mkTestbench_Or
compare_file Or.bsv.bsc-ccomp-out
}
#link_objects_pass {"mkTestbench_Or"} mkTestbench_Or
#sim_final_state mkTestbench_Or 10010
#compare_file "mkTestbench_Or.final-state"

test_c_veri_bsv_multi PrimSelectability mkTestbench_PrimSelectability {} "mkTestbench_PrimSelectability.out.expected" 

test_c_veri_bsv_multi Replicate mkTestbench_Replicate {} "mkTestbench_Replicate.out.expected" 

test_c_veri_bsv_multi Reverse mkTestbench_Reverse {} "mkTestbench_Reverse.out.expected" 

test_c_veri_bsv_multi Rotate mkTestbench_Rotate {} "mkTestbench_Rotate.out.expected" 

test_c_veri_bsv_multi RotateR mkTestbench_RotateR {} "mkTestbench_RotateR.out.expected" 

test_c_veri_bsv_multi Scanl mkTestbench_Scanl {} "mkTestbench_Scanl.out.expected" 

test_c_veri_bsv_multi Scanr mkTestbench_Scanr {} "mkTestbench_Scanr.out.expected" 

test_c_veri_bsv_multi Select mkTestbench_Select {} "mkTestbench_Select.out.expected" 

test_c_veri_bsv_multi Sscanl mkTestbench_Sscanl {} "mkTestbench_Sscanl.out.expected" 

test_c_veri_bsv_multi Sscanr mkTestbench_Sscanr {} "mkTestbench_Sscanr.out.expected" 

test_c_veri_bsv_multi Tail mkTestbench_Tail {} "mkTestbench_Tail.out.expected" 

test_c_veri_bsv_multi Take mkTestbench_Take {} "mkTestbench_Take.out.expected" 

test_c_veri_bsv_multi TakeWhile mkTestbench_TakeWhile {} "mkTestbench_TakeWhile.out.expected" 

test_c_veri_bsv_multi TakeWhileRev mkTestbench_TakeWhileRev {} "mkTestbench_TakeWhileRev.out.expected" 

test_c_veri_bsv_multi Transpose mkTestbench_Transpose {} "mkTestbench_Transpose.out.expected" 

test_c_veri_bsv_multi Unzip mkTestbench_Unzip {} "mkTestbench_Unzip.out.expected" 

test_c_veri_bsv_multi Update mkTestbench_Update {} "mkTestbench_Update.out.expected" 

test_c_veri_bsv_multi Upto mkTestbench_Upto {} "mkTestbench_Upto.out.expected" 

test_c_veri_bsv_multi Zip1 mkTestbench_Zip1 {} "mkTestbench_Zip1.out.expected" 

test_c_veri_bsv_multi Zip2 mkTestbench_Zip2 {} "mkTestbench_Zip2.out.expected" 

test_c_veri_bsv_multi Zip3 mkTestbench_Zip3 {} "mkTestbench_Zip3.out.expected" 

test_c_veri_bsv_multi Zip4 mkTestbench_Zip4 {} "mkTestbench_Zip4.out.expected" 

test_c_veri_bsv_multi ZipWith mkTestbench_ZipWith {} "mkTestbench_ZipWith.out.expected" 

test_c_veri_bsv_multi ZipWith3 mkTestbench_ZipWith3 {} "mkTestbench_ZipWith3.out.expected" 

test_c_veri_bsv_multi ZipWith4 mkTestbench_ZipWith4 {} "mkTestbench_ZipWith4.out.expected" 
