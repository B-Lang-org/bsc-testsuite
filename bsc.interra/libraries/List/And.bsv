package And;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_And();
   List #(Bool) my_list1 = Cons (True, Cons (True, Cons (True, Cons (True, Cons (True, Nil)))));
   List #(Bool) my_list2 = Cons (False, Cons (False, Cons (True, Cons (False, Cons (False, Nil)))));


   rule fire_once (True);
      $display("List1:");
      display_list (my_list1);
      $display("And Value = %d", and (my_list1));
      $display("List2:");
      display_list (my_list2);
      $display("And Value = %d", and (my_list2));
      
      if (and(my_list1) != True || and (my_list2) != False) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_And
endpackage : And
