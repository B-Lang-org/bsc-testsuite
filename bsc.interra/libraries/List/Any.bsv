package Any;

import List :: *;
function Bool f (Int #(32) a);
     if (a < 0)
        return True;
     else
        return False;
endfunction


module mkTestbench_Any();
   List #(Int #(32)) my_list1 = Cons (0, Cons (-1, Cons (2, Cons (-3, Cons (4, Nil)))));
   List #(Int #(32)) my_list2 = Cons (0, Cons (2, Cons (4, Nil)));



   rule fire_once (True);
      if (any(f, my_list1) != True || any(f, my_list2) != False) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Any
endpackage : Any
