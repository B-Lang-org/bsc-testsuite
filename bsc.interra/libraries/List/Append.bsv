//Appends to Lists of type UInt #(4).

package Append;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Append();
   List #(UInt #(4)) my_list1 = Cons (0, Cons (1, Cons (2, Cons (3, Cons (4, Nil)))));
   List #(UInt #(4)) my_list2 = Cons (5, Cons (6, Cons (7, Cons (8, Cons (9, Nil)))));
   List #(UInt #(4)) my_list3 = Cons (0, Cons (1, Cons (2, Cons (3, Cons (4, Cons (5, Cons (6, Cons (7, Cons (8, Cons (9, Nil))))))))));


   rule fire_once (True);
      $display("List1:");
      display_list (my_list1);
      $display("List2:");
      display_list (my_list2);
      $display("Appended List:");
      display_list (append(my_list1, my_list2));
      if (my_list3 != append (my_list1, my_list2)) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Append
endpackage : Append
