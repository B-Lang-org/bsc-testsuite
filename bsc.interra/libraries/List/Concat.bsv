// Concats four Lists and checks against the expected list
package Concat;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Concat();
   List #(Int #(32)) my_list1 = Cons (0, Cons (1, Cons (2, Cons (3, Cons (4, Nil)))));
   List #(Int #(32)) my_list2 = Cons (5, Cons (6, Cons (7, Cons (8, Cons (9, Nil)))));
   List #(Int #(32)) my_list3 = Cons (10, Cons (11, Cons (12, Cons (13, Cons (14, Nil)))));
   List #(Int #(32)) my_list4 = Cons (15, Cons (16, Cons (17, Cons (18, Cons (19, Nil)))));

   List #(List #(Int #(32))) my_list5 = Cons (my_list1, Cons (my_list2, Cons (my_list3, Cons (my_list4, Nil))));

   List #(Int #(32)) my_list6 = Cons(0, Cons (1, Cons (2, Cons (3, Cons (4, Cons (5, Cons (6, Cons (7, Cons (8, Cons (9, Cons (10, Cons (11, Cons (12, Cons (13, Cons(14, Cons (15, Cons (16, Cons (17, Cons (18, Cons (19, Nil))))))))))))))))))));


   rule fire_once (True);
      $display ("Concatenated List:");
      display_list (concat(my_list5));
      if (concat(my_list5) != my_list6) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Concat
endpackage : Concat
