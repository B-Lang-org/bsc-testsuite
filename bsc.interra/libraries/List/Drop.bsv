package Drop;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Drop();
   List #(Int #(4)) my_list1 = Cons (0, Cons (1, Cons (2, Cons (3, Cons (4, Nil)))));
   List #(Int #(4)) my_list2 = Cons (4, Nil);



   rule fire_once (True);
      $display ("Original List:");
      display_list (my_list1);
      $display ("Dropping four elements. New List:");
      display_list (drop(4, my_list1));
      if (drop(4, my_list1) != my_list2 || drop (6, my_list1) != Nil) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Drop
endpackage : Drop
