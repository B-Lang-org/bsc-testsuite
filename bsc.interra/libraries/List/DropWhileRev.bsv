package DropWhileRev;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction

function Bool f1 (Int #(4) a);
   if (a < 2)
       return True;
   else
       return False;
endfunction


function Bool f2 (Int #(4) a);
   return True;
endfunction

function Bool f3 (Int #(4) a);
   return False;
endfunction

module mkTestbench_DropWhileRev();
   List #(Int #(4)) my_list1 = Cons (0, Cons (1, Cons (2, Cons (0, Cons (1, Nil)))));
   List #(Int #(4)) my_list2 = Cons (0, Cons (1, Cons(2, Nil)));



   rule fire_once (True);
      $display ("Original List:");
      display_list (my_list1);
      $display ("Drops Tail Segment with elements < 2. New List:");
      display_list (dropWhileRev(f1, my_list1));
      if (dropWhileRev(f1, my_list1) != my_list2 || dropWhileRev (f2, my_list1) != Nil || dropWhileRev (f3, my_list1) != my_list1) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_DropWhileRev
endpackage : DropWhileRev
