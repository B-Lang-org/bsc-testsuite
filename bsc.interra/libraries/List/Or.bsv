package Or;

import List :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (List #(a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Or();
   List #(Bool) my_list1 = Cons (False, Cons (False, Cons (False, Cons (False, Cons (False, Nil)))));
   List #(Bool) my_list2 = Cons (False, Cons (False, Cons (True, Cons (False, Cons (False, Nil)))));


   rule fire_once (True);
      $display("List1:");
      display_list (my_list1);
      $display("Ored Value = %d", or (my_list1));
      $display("List2:");
      display_list (my_list2);
      $display("Ored Value = %d", or (my_list2));
      
      if (or(my_list1) != False || or (my_list2) != True) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
      $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Or
endpackage : Or
