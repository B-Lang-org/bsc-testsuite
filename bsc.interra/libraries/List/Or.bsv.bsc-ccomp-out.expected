checking package dependencies
Error: "Or.bsv", line 29, column 35: (P0005)
  Unexpected V2K keyword `or'; expected expression or `('
